### PR TITLE
Sync: Store and retrieve granular item count data

### PR DESCRIPTION
**Pull request by tyxla**
This updates our full sync module to store granular information about the item counts for each module, 
and adds that information to the set of status data that we pass to the `/sync/status` endpoint.

Part of #11985.

#### Changes proposed in this Pull Request:

* Now store the total sent items during full sync for each module in a new set of options.
* Pass the granular total sent items information to the sync status endpoint.
* Delete the new set of options when sync status is cleared.
* Amend the tests to include cases for the new functionality.

#### Testing instructions:

* Apply D27227-code and follow the test instructions there.

#### Proposed changelog entry for your changes:

* Now storing, maintaining and retrieving granular module item count information as part of the sync status endpoint.

### CODE REVIEW

File: sync/class.jetpack-sync-module-full-sync.php [hunk 1]
```diff
@@ -242,17 +244,27 @@ function update_sent_progress_action( $actions ) {
...
+				if ( ! empty( $actions_with_total_counts[ $module_action ] ) ) {
+					$items_sent_total += $actions_with_total_counts[ $module_action ];
+				}
 			}
 
 			if ( $items_sent > 0 ) {
 				$this->update_status_option( $status_option_name, $items_sent );
 			}
+
+			if ( $items_sent_total !== 0 ) {
```
Comment (kraftbj) [inline sync/class.jetpack-sync-module-full-sync.php]:
```suggestion
			if ( 0 !== $items_sent_total ) {
```

File: sync/class.jetpack-sync-module-full-sync.php [hunk 2]
```diff
@@ -267,6 +279,33 @@ public function get_action_name( $queue_item ) {
 		return false;
 	}
 
+	public function get_action_totals( $queue_item ) {
```
Comment (kraftbj) [inline sync/class.jetpack-sync-module-full-sync.php]: Could you add a quick docblock for this function and the other new one added? (I'm trying to slowly get everything to coding standards as things are touched in existing files)

File: sync/class.jetpack-sync-module-full-sync.php [hunk 3]
```diff
@@ -267,6 +279,33 @@ public function get_action_name( $queue_item ) {
...
+			return -1;
+		}
+		return 0;
+	}
+
+	public function get_actions_totals( $actions ) {
+		$totals = array();
+
+		foreach ( $actions as $action ) {
+			$name = $this->get_action_name( $action );
```
Comment (kraftbj) [inline sync/class.jetpack-sync-module-full-sync.php]:
```suggestion
			$name          = $this->get_action_name( $action );
```

File: sync/class.jetpack-sync-module-full-sync.php [hunk 4]
```diff
@@ -309,6 +349,10 @@ public function get_status() {
 			if ( $sent = $this->get_status_option( "{$name}_sent" ) ) {
 				$status['sent'][ $name ] = $sent;
 			}
+
+			if ( $sent_total = $this->get_status_option( "{$name}_sent_total" ) ) {
```
Comment (kraftbj) [inline sync/class.jetpack-sync-module-full-sync.php]:
```suggestion
			$sent_total = $this->get_status_option( "{$name}_sent_total" );
			if ( $sent_total ) {
```

File: sync/class.jetpack-sync-module-full-sync.php [hunk 5]
```diff
@@ -267,6 +279,33 @@ public function get_action_name( $queue_item ) {
 		return false;
 	}
 
+	public function get_action_totals( $queue_item ) {
+		if ( is_array( $queue_item ) && isset( $queue_item[1] ) && isset( $queue_item[1][0] ) ) {
```
Comment (kraftbj) [inline sync/class.jetpack-sync-module-full-sync.php]: This is more a general comment, but it's really hard to parse what `[1][0]` is supposed to be. Maybe detail it a bit in the docblock or reference where this is documented?

**Review by kraftbj — DISMISSED** [review]
A couple of minor things and documentation request. Would like a Poseidon 👍 before merging after it has been updated.

File: tests/php/sync/test_class.jetpack-sync-full.php [hunk 6]
```diff
@@ -11,6 +11,9 @@ class WP_Test_Jetpack_Sync_Full extends WP_Test_Jetpack_Sync_Base {
 	private $full_sync_start_config;
 	private $synced_user_ids;
 
+	private $test_posts_count = 20;
```
Comment (aldavigdis) [inline tests/php/sync/test_class.jetpack-sync-full.php]: Just a nit-pick, but aligning the `=` with the line below would be a good idea.

File: sync/class.jetpack-sync-module-full-sync.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (aldavigdis) [reply sync/class.jetpack-sync-module-full-sync.php]: If so, then perhaps using the opportunity and fixing the statement above to be a Yoda condition as well is a good idea?

**Review by aldavigdis — DISMISSED** [review]

File: sync/class.jetpack-sync-module-full-sync.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (kraftbj) [reply sync/class.jetpack-sync-module-full-sync.php]: That one does need to be ( https://github.com/Automattic/jetpack/pull/12125#discussion_r277715870 ), but the one above it doesn't since there's no way to accidentally assign `$items_sent_total` a value.

File: sync/class.jetpack-sync-module-full-sync.php [hunk 7]
```diff
@@ -242,17 +244,27 @@ function update_sent_progress_action( $actions ) {
 		foreach ( Jetpack_Sync_Modules::get_modules() as $module ) {
 			$module_actions     = $module->get_full_sync_actions();
 			$status_option_name = "{$module->name()}_sent";
+			$total_option_name  = "{$status_option_name}_total";
 			$items_sent         = $this->get_status_option( $status_option_name, 0 );
```
Comment (lezama) [inline sync/class.jetpack-sync-module-full-sync.php]: not a request and probably the safest approach is to keep both, but I wonder if we could get rid of "items_sent"

File: sync/class.jetpack-sync-module-full-sync.php (hunk 2 above)
Comment (tyxla) [reply sync/class.jetpack-sync-module-full-sync.php]: Great call, thanks! Added docblocks for the new methods here: 8641517ec3ea434912663488682aef82f79310a8

File: sync/class.jetpack-sync-module-full-sync.php (hunk 5 above)
Comment (tyxla) [reply sync/class.jetpack-sync-module-full-sync.php]: I understand the confusion - added some information in the docblock that came with  8641517ec3ea434912663488682aef82f79310a8

File: sync/class.jetpack-sync-module-full-sync.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (tyxla) [reply sync/class.jetpack-sync-module-full-sync.php]: I never got used to using Yoda conditions, and TBH I'm not a huge fan of them 😅 

Updated though, thanks. 😉

File: sync/class.jetpack-sync-module-full-sync.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (tyxla) [reply sync/class.jetpack-sync-module-full-sync.php]: Great call! In 1dddb3e61ecf5e0c58ba359e8795d1931d277d42 I've left a single `isset()` as there is no need for two of them.

File: sync/class.jetpack-sync-module-full-sync.php (hunk 7 above)
Comment (tyxla) [reply sync/class.jetpack-sync-module-full-sync.php]: I'm happy to remove it, but also I'm not sure if we use it somewhere else. It would be hard to verify it's no longer used anywhere - maybe it's best to keep it for now, and consider cleaning up in the future.

**Comment by tyxla** [comment]
Thanks for the awesome feedback @aldavigdis @kraftbj and @lezama! This is ready for another review.

File: sync/class.jetpack-sync-module-full-sync.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (aldavigdis) [reply sync/class.jetpack-sync-module-full-sync.php]: Not a big fan myself, just thought this would be useful. :)

**Review by aldavigdis — APPROVED** [review]
Very nice. I like.

**Review by lezama — APPROVED** [review]
🚢 IT

**Comment by tyxla** [comment]
Thank you, folks 🙌
